test/distributed/org/apache/cassandra/distributed/test/ShortReadProtectionTest.java

```diff
@@ -66,51 +329,192 @@ public void testGroupBySRPRegularRow() throws Throwable
+                         "INSERT INTO %s (pk, ck) VALUES (2, 1) USING TIMESTAMP 42",
+                         "INSERT INTO %s (pk, ck) VALUES (4, 0) USING TIMESTAMP 42",
+                         "INSERT INTO %s (pk, ck) VALUES (4, 1) USING TIMESTAMP 42")
+              .assertRows("SELECT pk, ck FROM %s PER PARTITION LIMIT 2 LIMIT 4",
+                          row(0, 0), row(0, 1), row(4, 0), row(4, 1));
+    }
+
+    private static long token(int key)
+    {
+        return (long) Murmur3Partitioner.instance.getToken(ByteBufferUtil.bytes(key)).getTokenValue();
+    }
+
+    private static class Tester
+    {
+        private static final AtomicInteger seqNumber = new AtomicInteger();
+
+        private final ICoordinator coordinator;
+        private final boolean quorum, flush, paging;
+        private final String qualifiedTableName;
+        private final ConsistencyLevel consistencyLevel;
+
+        private boolean flushed = false;
+
+        private Tester(boolean quorum, boolean flush, boolean paging, int coordinator)
+        {
+            this.coordinator = cluster.coordinator(coordinator);
+            this.quorum = quorum;
+            this.flush = flush;
+            this.paging = paging;
+            qualifiedTableName = KEYSPACE + ".t_" + seqNumber.getAndIncrement();
+            consistencyLevel = quorum ? QUORUM : ALL;
+        }
+
+        private Tester createTable(String query)
+        {
+            cluster.schemaChange(format(query) + " WITH read_repair='NONE'");
+            return this;
+        }
+
+        private Tester allNodes(int startInclusive, int endExclusive, Function<Integer, String> querySupplier)
+        {
+            IntStream.range(startInclusive, endExclusive).mapToObj(querySupplier::apply).forEach(this::allNodes);
+            return this;
+        }
+
+        private Tester allNodes(String... queries)
+        {
+            for (String query : queries)
+                allNodes(query);
+            return this;
+        }
+
+        private Tester allNodes(String query)
+        {
+            coordinator.execute(format(query), ALL);
+            return this;
+        }
+
+        private Tester onlyNode1(String... queries)
+        {
+            onlyNodeN(1, queries);
+            return this;
+        }
+
+        private Tester onlyNode2(String... queries)
+        {
+            onlyNodeN(2, queries);
+            return this;
+        }
+
+        private Tester onlyNode3(String... queries)
+        {
+            onlyNodeN(3, queries);
+            return this;
+        }
+
+        private void onlyNodeN(int node, String... queries)
+        {
+            for (String query : queries)
+            {
+                String formattedQuery = format(query);
+                cluster.get(node).executeInternal(formattedQuery);
+                if (quorum)
+                    cluster.get(node == NUM_NODES ? 1 : node + 1).executeInternal(formattedQuery);
```

Comment:
@adelapena This confused me a bit when I first saw it. (i.e. `onlyNode1()` sometimes actually doesn't write just to node 1.) The reason we have these helpers is to make the higher level tests more intelligible, but I'm afraid this is a bit misleading. Perhaps some simple renaming would be fine, something like `toNodeN()` and a bit of documentation? I'll think it over and see if I can come up with any other ideas...

Reply:
This brings me to another point. If we're really focused on coordination here, perhaps we could just use `executeInternal()` for all writes, make things uniform. It isolates us from bugs in replication I, but perhaps I'm being too picky...ignore if you don't think it matters.

Reply:
I think all writes are already internal. I have renamed the methods and added some documentation in an attempt to make the relationship between writes and reads clearer, please let me know what do you think.